config/auth_metadata_service.go

```diff
@@ -0,0 +1,166 @@
+package config
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/databricks/databricks-sdk-go/logger"
+	"golang.org/x/oauth2"
+)
+
+// timeout to wait for metadata service to respond
+const metadataServiceTimeout = 10 * time.Second
+
+const MetadataServiceVersion = "1"
+
+// response expected from the metadata service
+// other fields such as `cluster_id` might be added later
+type serverResponse struct {
+	// URL of the Databricks workspace to connect to
+	Host string `json:"host"`
+
+	// The requested access token.
+	Token oauth2.Token `json:"token"`
+}
+
+// Credentials provider that fetches a token from a locally running HTTP server
+//
+// The credentials provider will perform a GET request to the configured URL.
+//
+// The MUST return 4xx response if the "X-Databricks-Metadata-Version" header
+// is not set or set to a version that the server doesn't support.
+//
+// The server MUST guarantee stable sessions per URL path. That is, if the
+// server returns a token for a Host on a given URL path, it MUST continue to return
+// tokens for the same Host.
+//
+// The server is expected to return a JSON response with the following fields:
+//
+// host: URL of the Databricks host to connect to.
+// token: object with the following fields
+//   access_token: The requested access token.
+//	 token_type: The type of token, which is a "Bearer" access token.
+//	 expires_on: The timespan when the access token expires.
+
+type MetadataServiceCredentials struct{}
+
+func (c MetadataServiceCredentials) Name() string {
+	return "metadata-service"
+}
+
+func (c MetadataServiceCredentials) Configure(ctx context.Context, cfg *Config) (func(*http.Request) error, error) {
+	if cfg.MetadataServiceURL == "" {
+		return nil, nil
+	}
+
+	parsedMetadataServiceURL, err := url.Parse(cfg.MetadataServiceURL)
```

Comment:
please make sure there's a reuse between Azure MSI and this and there's compatibility. that's the main requirement for this PR to pass, otherwise it explodes the complexity for all SDKs.

Reply:
I've looked at the diff between the two implementations. The two are similar but not similar enough to share implementation without increasing complexity.

The only way to meaningfully unify would be to use exactly the MSI protocol and wire format but that prevents us evolving the protocol e.g. adding `host` and `cluster_id` to the payload. Even if we do that only the ~40 lines of the `Token` function can be re-used https://github.com/databricks/databricks-sdk-go/blob/main/config/auth_azure_msi.go#LL97-L136C2

Reply:
can you extend/reuse the `azureMsiTokenSource` - https://github.com/databricks/databricks-sdk-go/blob/main/config/auth_azure_msi.go#L104-L157. You can easily do that by passing down `func(*http.Request) error` callback, which will do something like:

```
customizeMetadata := func (req *http.Request) error {
    req.Header.Add(MetadataServiceVersionHeader, MetadataServiceVersion)
    req.Header.Add(MetadataServiceHostHeader, s.config.Host)
   return nil
}

customizeMsi := func (req *http.Request) error {
    query := req.URL.Query()
	query.Add("api-version", "2018-02-01")
	query.Add("resource", s.resource)
	if s.clientId != "" {
		query.Add("client_id", s.clientId)
	}
	req.URL.RawQuery = query.Encode()
	req.Header.Add("Metadata", "true")
    return nil
}
```